Automattic/wp-calypso
# Masterbar - always show site items + add prevSelectedSiteId to calypso state

Related to # p1721229953008389-slack-C06DN6QQVAQ

## Proposed Changes

Allows site specific items to remain in the masterbar at global levels where selectedSite is cleared for various reasons.

* Adds a `prevSelectedSiteId` to the calypso redux store. We do not need an action for this, we just set dispatches to it when calls for `setSelectedSiteId` or `setAllSitesSelected` are called. `prevSelectedSiteId` may be null initially, but we prevent setting null values on changes to always have context of the most previously selected site.
    * Updates and adds unit tests related to these actions, reducer, and selectors. 
    * Updated 2 `useDispatch` imports in stepper files that were using this selector with the useDispatch directly from react-redux. The types from react-redux do not seem to support thunks, so we import it from calypso/state instead which types the redux dispatch to support thunks.
* Updates the masterbar to always try to use a site. The priority goes from `currentlySelectedSite` > `previouslySelectedSite` > `primarySite`. This effects the site Menu, +new menu, and checkout cart.
    *  We may want to rethink or separate conditions for when we show which items for which sites. But this seems like a good starting point to try.

BEFORE
![prev-site-before](https://github.com/user-attachments/assets/3adaf427-3917-4c0c-acf0-31109c701fb3)

AFTER
![prev-site-after](https://github.com/user-attachments/assets/fefb6c40-fc29-470b-a91f-72ee70dbc91b)

## Why are these changes being made?

 Context from p1721229953008389-slack-C06DN6QQVAQ
* We want the global masterbar to be consistent while navigating throughout global pages. The domains, themes, plugins, etc. at global levels are tangled with the selectedSiteId - deselecting it on load to accommodate the global context and usage of these pages.
* Adding a prevSelectedSiteId and applying these changes to state seems like the safest way to ensure continuity when navigating around calypso global levels, and these priorities for siteId in the masterbar allow for it to always have context for a site if one exists.

## Testing Instructions



* Go to calypso /sites and select a site to overview.
* verify the site menu, +new,  etc, links continue work as expected.
* close the overview tab, verify the site menu does not disappear.
* verify the site menu is still present (previously would disappear at this point) and that, +new, etc. links continue to work as expected.
* click into the domains tab on the global sidebar. Verify the site specific items are still in the global sidebar. verify the masterbar and this domains page work as expected.
* do the above for global plugins and themes pages.
* smoke test around calypso. verify things work as expected and we haven't created and weird edge cases.

## Pre-merge Checklist
